spacy/tokenizer.pyx

```diff
@@ -174,23 +211,124 @@ cdef class Tokenizer:
+        # If modifications never increase doc length, can modify in place
+        if modify_in_place:
+            tokens = doc.c
+        # Otherwise create a separate array to store modified tokens
+        else:
+            tokens = <TokenC*>mem.alloc(max_length, sizeof(TokenC))
+        # Modify tokenization according to filtered special cases
+        i = 0
+        while i < doc.length:
+            if not i in span_data:
+                tokens[i + offset] = doc.c[i]
+                i += 1
+            else:
+                span = span_data[i]
+                cached = <_Cached*>self._specials.get(hash_string(span[0]))
+                if cached == NULL:
+                    # Copy original tokens if no rule found
+                    for j in range(span[2] - span[1]):
+                        tokens[i + offset + j] = doc.c[i + j]
+                    i += span[2] - span[1]
+                else:
+                    # Copy special case tokens into doc and adjust token and
+                    # character offsets
+                    idx_offset = 0
+                    orig_final_spacy = doc.c[span[2] + offset - 1].spacy
+                    orig_idx = doc.c[i].idx
+                    for j in range(cached.length):
+                        tokens[i + offset + j] = cached.data.tokens[j]
+                        tokens[i + offset + j].idx = orig_idx + idx_offset
+                        idx_offset += cached.data.tokens[j].lex.length + \
+                                1 if cached.data.tokens[j].spacy else 0
+                    tokens[i + offset + cached.length - 1].spacy = orig_final_spacy
+                    i += span[2] - span[1]
+                    offset += span[3]
+        # Allocate more memory for doc if needed
+        cdef int modified_doc_length = doc.length + offset
+        while modified_doc_length >= doc.max_length:
+            doc._realloc(doc.length * 2)
+        # If not modified in place, copy tokens back to doc
+        if not modify_in_place:
+            memcpy(doc.c, tokens, max_length * sizeof(TokenC))
+        for i in range(doc.length + offset, doc.length):
+            memset(&doc.c[i], 0, sizeof(TokenC))
+            doc.c[i].lex = &EMPTY_LEXEME
+        doc.length = doc.length + offset
+        return True
+
+    cdef void _filter_spans(self, vector[MatchStruct] &original, vector[MatchStruct] &filtered, int doc_len) nogil:
+
+        cdef int seen_i
+        cdef MatchStruct span
+        cdef stdset[int] seen_tokens
+        stdsort(original.begin(), original.end(), len_start_cmp)
+        cdef int orig_i = original.size() - 1
+        while orig_i >= 0:
+            span = original[orig_i]
+            if not seen_tokens.count(span.start) and not seen_tokens.count(span.end - 1):
+                filtered.push_back(span)
+            for seen_i in range(span.start, span.end):
+                seen_tokens.insert(seen_i)
+            orig_i -= 1
```

Comment:
So if I understand correctly, this matches the biggest spans first, and if two spans are equally big, it gives preference to the **last** one? Because with equal length you're sorting on `a.start < b.start` but you're processing the list backwards. Just checking whether this was the intended behaviour. Personally I think I would have given preference to the first span in the `doc` (when equal length), but there's probably not a "best" answer, so maybe it doesn't matter either way.

Reply:
The idea is that this function is identical to `util.filter_spans()`. We should add the same tests for this function, too.

I have to think about whether it's doing the right thing in those cases. Some of the details are simpler because it's not possible to have two spans over the same tokens. (There's only one special rule for any given token sequence.)

Reply:
You're right that it prefers the later longer span rather than the earlier one. If we change this here, we should also change `util.filter_spans()`. I don't think that this comes up a lot in practice with the tokenizer because the matches tend to be short and not overlapping (one real case might be `'''` becomes `' ''` rather than `'' '`), but it is relevant for `util.filter_spans()`, which is used for filtering things like entities and doesn't do exactly what it says, either.

Reply:
Ah, you're right. So we should either change the sorting behaviour or the comment, at least. Any thoughts, @honnibal ? (https://github.com/explosion/spaCy/blob/master/spacy/util.py#L629 sorts on size small->big and start first->later, but then that order is reversed)

Reply:
I think we should change the behavior for both to prefer the earliest span, as seems to be intended from the documentation. I think we can just swap the compare here and change `span.start` to `-span.start` in the other? I think @ines is the original author?

Reply:
@adrianeboyd Yes, good point, I hadn't realised that the behaviour didn't actually match the documentation. (Sorry about that.) So yes, let's change that. 

Also, if there's anything we can combine here to avoid code duplication, that'd be good, too. I've also always wanted a more powerful `filter_spans` helper that implements different ways of filtering (prefer longest, prefer shortest, maybe prefer as many as possible?).

Reply:
I didn't think about this problem when I was reimplementing it, either!

I originally used the method from `util` and only added this one to try to make a little faster. It would obviously be much nicer to just maintain one method, hmm. I'm using the `MatchStruct` struct, but they're really just spans underneath and there's no need for it to be part of `PhraseMatcher`.

I suppose we could create a top-level `SpanC` kind of struct instead of `MatchStruct` and move the `util` method to `Span` or `Doc`? The `util` method doesn't prevent you passing in spans from different docs, either, so you can get really weird output if you mix things up and that might be nice to check or prevent.

Reply:
`Entity` is already there. Does it make sense to use it here, too? Rename it, though, maybe?